src/osd/PGBackend.cc

```diff
@@ -1001,11 +1001,12 @@ void PGBackend::be_compare_scrubmaps(
 	  if (found)
 	    errorstream << pgid << " shard " << j->first << ": soid " << *k
 		      << " " << ss.str() << "\n";
-	} else if (found) {
+	} else if (object_error.errors != 0) {
```

Comment:
my concern is that, we don't reset `object_error` before moving to another `ScrubMap`. so once we identify an error in a certain shard of current object, we will always go to this branch if there is no shard error in succeeding shards. but the reason we go to this branch is not necessary that the `object_error.errors` is contributed by current shard.

Reply:
@tchaikov Good catch, but will require a follow-on fix.  I now realize that we had a bug when I introduced some inconsistency checking in `be_select_auth_object()`.  This change actually improves repair of those inconsistencies at the expense of repairing all shards instead of just the ones that don't match the authoritative shard.  This is a result of not resetting `object_error.errors`.

I had already filed a tracker http://tracker.ceph.com/issues/25108 to cover a logging issue with errors found in `be_select_auth_object()`.  But really I need to move all error checking out of `be_select_auth_object()` and into `be_compare_scrub_objects()`.  This will fix the logging issue.  Then we can return to `} else if (found) {` which is only needed for logging.  At that point the `set<pg_shard_t> object_errors` can go away **if** `auth_list` can't be empty.   I'm not totally sure about that yet.

Reply:
@tchaikov I've got a first cut at restructuring of this code to address your concern and tracker 25108.